### PR TITLE
Implement Long and Integer's parseUnsigned and toStringUnsigned

### PR DESCRIPTION
**Pull request by pbatko**
See #242

### CODE REVIEW

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 1]
```diff
@@ -451,9 +451,80 @@ object Integer {
...
+    while (offset < length) {
+      val digit = Character.digit(s.charAt(offset), radix)
+      offset += 1
+
+      if (digit == -1) throw new NumberFormatException(s)
+
+      if (compareUnsigned(result, max) == 1) throw new NumberFormatException(s)
+
+      result = result * radix + digit
+    }
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/Integer.scala]: One test case for you: `Integer.parseUnsignedInt("4294967296")`. It should fail with `NumberFormatException`, but currently this will succeed and return 0. That's because the (unsigned) addition overflows.
You need to test that, after the addition of `digit`, `result` is not smaller than `digit` (which would signal an overflow during the addition):
```scala
if (compareUnsigned(result, digit) < 0) throw new NumberFormatException(s)
```

File: javalib/src/main/scala/java/lang/Long.scala [hunk 2]
```diff
@@ -338,9 +338,82 @@ object Long {
...
+    while (offset < length) {
+      val digit = Character.digit(s.charAt(offset), radix)
+      offset += 1
+
+      if (digit == -1) throw new NumberFormatException(s)
+
+      if (compareUnsigned(result, max) == 1) throw new NumberFormatException(s)
+
+      result = result * radix + digit
+    }
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/Long.scala]: The same problem with addition overflow appears here. Test case: `JLong.parseUnsignedLong("18446744073709551616")`

**Comment by pbatko** [comment]
@sjrd I added the check, but not in the exact form you suggested. Can you take a look?

File: javalib/src/main/scala/java/lang/Integer.scala [hunk 3]
```diff
@@ -483,9 +482,11 @@ object Integer {
 
       if (digit == -1) throw new NumberFormatException(s)
 
-      if (compareUnsigned(result, max) == 1) throw new NumberFormatException(s)
+      val next = result * radix + digit
 
-      result = result * radix + digit
+      if (compareUnsigned(next, result) < 0) throw new NumberFormatException(s)
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/Integer.scala]: You cannot use comparison after the fact to check for overflow of *multiplication*. That's because multiplication overflow does not "wrap around" in any meaningful way. It just overflows and bits are lost.

For example, in 8 bits, consider `80 * 8`. 80 is `01010000` in binary. If `result == 80`, `radix == 8` and `digit == 0` at this point, then `next = result * radix + digit == 80 * 8 + 0 == 80 << 3 == 01010000b << 3 == 10000000b == 128`. You have `next` is unsigned-greater than `result`, but you had an overflow.

The previous check with `max` that you had was the good way to test for overflow of the *multiplication*. My addition is the good way to test for overflow of the *addition*. You need both.

**Comment by densh** [comment]
Please squash and rebase on top of latest master.

**Comment by densh** [comment]
LGTM, thanks @pbatko ! 👍
